examples/auth/cadc_auth_example.py

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Example for authenticating with CADC TAP service
+"""
+import getpass
+
+import requests
+
+import pyvo
+import pyvo.extensions.auth.authsession
+
+# Gather login information
+data = {
+    'username': input('Username:'),
+    'password': getpass.getpass('Password:')
+}
+
+headers = {
+    'Content-Type': 'application/x-www-form-urlencoded',
+    'Accept': 'text/plain'
+}
+
+# Do the login and get the cookie
+login_url = 'https://ws-cadc.canfar.net/ac/login'
+response = requests.post(login_url, data=data, headers=headers)
+response.raise_for_status()
+cookie = '\"' + response.text + '\"'
+
+# Configure the session and run the query
+auth = pyvo.extensions.auth.authsession.AuthSession()
+auth.credentials.set_cookie('CADC_SSO', cookie)
+service = pyvo.dal.TAPService('https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/tap', auth)
```

Comment:
Well I can make the import names shorter, or just put it in the global namespace if you want.  I think that's reasonable.

As I believe I mentioned on the previous PR and in our slack conversations, I very much disagree with the concept of passing a dictionary in for this as opposed to an object, which I will detail in a much longer later comment.

Reply:
I don't have a strong opinion on the dictionary except that in Python it seems to be the defacto way of passing data structures. The `CredentialStore` class has its own advantages as you've list them so I'm fine with that (in `opencadc` we also use a class to pass user credentials). I would still however build the `AuthSession` inside the `TAPService` constructor unless there is a good reason to leave it on the user to do that. The code could become something like:
```
from pyvo.extensions import CredentialStore
cred = CredentialStore()
cred.set_cookie(('name', 'value'))  # or cred.set_cookie(cookie_jar)
cred.set_client_certificate('certificate_path')
cred.set_auth_session('key', session)
tap = pyvo.dal.TAPService('https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/tap', auth)
```

Reply:
So the only thing I see as different from what your code is and what I'm doing in the example is that it gets rid of calling into a member of the subclass of authsession.  If I flatten this out (put all the credstore logic into authsession, and then call it credstore) would that be fine?  I think that's the only difference I can see.

The reason I don't want to compose them inside of the TAP service is that then I would need to compose it the same way in every subobject.  For example, TAPService makes objects like AsyncTAPResults and TAPResults which use the sessions.  What they need is the creds, and the mapping (can only be gotten from tap service).  So TAPService needs to pass all this info down.  It currently passes it as one object, the authsession.

Composing the object everywhere it is used is somewhat trouble, since I don't have all the info required to make it.  So every child helper class would need to take an authsession.  In this way, at least it's consistent that every class takes a session.

In the way I've done it, there are theoretically very limited changes for people who aren't using auth (they are just not calling a global session anymore).  Then we simply pass in a class that also works as a session.  Someone could even pass in something other than authsession, like a different requests.Session entirely if they wanted to.  In this way, the classes like TAPService and all its child classes don't need to do anything to session, they just call it (either GET or POST, that's it).